cmd/soroban-rpc/internal/xdr2json/conversion.go

```diff
@@ -0,0 +1,166 @@
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"unsafe"
+
+	"github.com/pkg/errors"
+
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/db"
+)
+
+const (
+	FormatBase64 = "base64"
+	FormatJSON   = "json"
+)
+
+var errInvalidFormat = fmt.Errorf(
+	"expected %s for optional 'xdrFormat'",
+	strings.Join([]string{FormatBase64, FormatJSON}, ", "))
+
+// Convert takes an XDR object (`xdr`) and its serialized bytes (`field`)
+// and returns the JSON-formatted serialization of that object.
+//
+// The `xdr` object does not need to actually be initialized/valid:
+// we only use it to determine the name of the structure. We could just
+// accept a string, but that would make mistakes likelier than passing the
+// structure itself (by reference).
+func Convert(xdr interface{}, field []byte) (map[string]interface{}, error) {
+	xdrTypeName := reflect.TypeOf(xdr).Name()
+	goStr := convertStr(xdrTypeName, field)
+	return jsonify(goStr)
+}
+
+// ConvertAny takes a valid XDR object (`xdr`) and returns a
+// JSON-formatted serialization of that object.
+//
+// Unlike `Convert`, the value here needs to be valid and
+// serializable.
+func ConvertAny(xdr interface{}) (map[string]interface{}, error) {
+	jsonStr, err := convertAnyStr(xdr)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonify(jsonStr)
+}
+
+func convertAnyStr(xdr interface{}) (string, error) {
+	xdrTypeName := reflect.TypeOf(xdr).Name()
+	if cerealXdr, ok := xdr.(encoding.BinaryMarshaler); !ok {
+		data, err := cerealXdr.MarshalBinary()
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to serialize XDR type '%s'", xdrTypeName)
+		}
+
+		return convertStr(xdrTypeName, data), nil
+	}
+
+	return "", fmt.Errorf("expected serializable XDR, got '%s': %+v", xdrTypeName, xdr)
+}
+
+func convertStr(xdrTypeName string, field []byte) string {
+	var goStr string
+	// scope just added to show matching alloc/frees
+	{
+		goRawXdr := CXDR(field)
+		b := C.CString(xdrTypeName)
+
+		result := C.xdr_to_json(b, goRawXdr)
+		C.free(unsafe.Pointer(b))
+
+		goStr = C.GoString(result)
+		C.free(unsafe.Pointer(result))
+	}
+
+	return goStr
+}
+
+func jsonify(s string) (map[string]interface{}, error) {
+	var result map[string]interface{}
+	err := json.Unmarshal([]byte(s), &result)
+	if err != nil {
+		return nil, err
+	} else if jsonErr, ok := result["error"]; ok {
+		return nil, fmt.Errorf("error during conversion: %+v", jsonErr)
+	}
+
+	return result, nil
+}
+
+// CXDR is ripped directly from preflight.go.
+func CXDR(xdr []byte) C.xdr_t {
+	return C.xdr_t{
+		xdr: (*C.uchar)(C.CBytes(xdr)),
+		len: C.size_t(len(xdr)),
+	}
+}
+
+func TransactionToJSON(tx db.Transaction) (
+	map[string]interface{},
+	map[string]interface{},
+	map[string]interface{},
+	[]map[string]interface{},
+	error,
+) {
+	var err error
+	var result, envelope, resultMeta map[string]interface{}
+	var diagEvents []map[string]interface{}
+
+	result, err = Convert(xdr.TransactionResult{}, tx.Result)
```

Comment:
Reviewer note: this provides type safety over passing e.g., `"TransactionResult"` which is prone to typos